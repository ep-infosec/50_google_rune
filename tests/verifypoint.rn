//  Copyright 2021 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Verify (gx, gy) is on the curve: -x**2 + y**2 = 1 + dx**2y**2:
func verifyPoint(p, d, m) {
  x = p[0]; y = p[1]
  println "x = ", x, ", y = ", y
  if -x**2 + y**2 != 1 + d*x**2*y**2 mod m {
    throw "Point not on curve"
  }
}

// Curve 25519
m = <u255>(2u256**255 - 19u256)
println "modulus = ", m
d = -121665/121666 mod m
println "d = ", d
// The generator's Y coordinate.
gy = 4 / 5 mod m
// Computing this requires modular square root.
gx = 15112221349535400772501151409588531511454012693041857206046113283949847762202u255
println "gy * 5 = ", gy * 5 mod m
g = (gx, gy)
verifyPoint(g, d, m)
