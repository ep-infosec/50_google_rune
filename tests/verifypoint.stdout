modulus = 57896044618658097711785492504343953926634992332820282019728792003956564819949
d = 37095705934669439343138083508754565189542113879843219016388785533085940283555
gy * 5 = 4
x = 15112221349535400772501151409588531511454012693041857206046113283949847762202, y = 46316835694926478169428394003475163141307993866256225615783033603165251855960
